### PR TITLE
mimic: osd: race condition opening heartbeat connection

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/36637

### CODE REVIEW

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -4756,9 +4756,9 @@ void OSD::heartbeat()
 
 bool OSD::heartbeat_reset(Connection *con)
 {
+  std::lock_guard l(heartbeat_lock);
```
Comment (smithfarm) [inline src/osd/OSD.cc]: According to https://en.cppreference.com/w/cpp/thread/lock_guard `std::lock_guard` was introduced in C++17?

I guess this won't work in mimic, and definitely not in luminous.

Comment (liewegas) [reply src/osd/OSD.cc]: use Mutex::Locker instead

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -4756,9 +4756,9 @@ void OSD::heartbeat()
 
 bool OSD::heartbeat_reset(Connection *con)
 {
+  Mutex::Locker l(heartbeat_lock);
   auto s = con->get_priv();
   if (s) {
-    heartbeat_lock.Lock();
     if (is_stopping()) {
       heartbeat_lock.Unlock();
```
Comment (gregsfortytwo) [inline src/osd/OSD.cc]: This backport is missing a follow-on patch that fixed up this erroneous extra unlock.

**Comment by yuriw** [comment]
[wip-yuri5-testing-2018-11-16-0044-mimic](https://trello.com/c/HPawazs3/474-wip-yuri5-testing-2018-11-16-0044-mimic)

**Pull request merged**